File: server-rs/src/settings/network_settings.rs
use super::lan::LanManager;
use super::network_config::{
    config_differs_from_running, hosts_file_line, load_network_config, save_network_config,
    NetworkConfig, NetworkConfigPatch,
};
use crate::types::NetworkSettings;

pub fn get_running_network_config(lan: &LanManager) -> NetworkConfig {
    NetworkConfig {
        host: lan.get_local_host().to_string(),
        port: lan.port,
    }
}

pub fn build_network_settings(lan: &LanManager) -> NetworkSettings {
    let running = get_running_network_config(lan);
    let saved = load_network_config();
    NetworkSettings {
        host: saved.host.clone(),
        port: saved.port,
        local_url: lan.get_local_url(),
        scheme: lan.get_url_scheme().to_string(),
        hosts_file_line: hosts_file_line(&saved.host),
        pending_restart: config_differs_from_running(&saved, &running),
    }
}

pub fn patch_network_settings(
    lan: &LanManager,
    patch: &NetworkConfigPatch,
) -> crate::error::AppResult<NetworkSettings> {
    save_network_config(patch)?;
    Ok(build_network_settings(lan))
}

File: server/modules/settings/network-settings.ts
import type { NetworkSettings } from "../../schemas/api.js";
import type { LanManager } from "./lan.js";
import {
  configDiffersFromRunning,
  hostsFileLine,
  loadNetworkConfig,
  nonProdPort,
  saveNetworkConfig,
  type NetworkConfig,
  type PatchNetworkConfig,
} from "./network-config.js";

export function getRunningNetworkConfig(lan: LanManager): NetworkConfig {
  return { host: lan.getLocalHost(), port: lan.port };
}

export function buildNetworkSettings(lan: LanManager): NetworkSettings {
  const running = getRunningNetworkConfig(lan);
  const saved = loadNetworkConfig();
  return {
    host: saved.host,
    port: saved.port,
    prodPort: saved.port,
    nonProdPort: nonProdPort(saved.port),
    localUrl: lan.getLocalUrl(),
    scheme: lan.getUrlScheme(),
    hostsFileLine: hostsFileLine(saved.host),
    pendingRestart: configDiffersFromRunning(saved, running),
  };
}

export function patchNetworkSettings(
  lan: LanManager,
  patch: PatchNetworkConfig,
): NetworkSettings {
  saveNetworkConfig(patch);
  return buildNetworkSettings(lan);
}
